firmware/library/L1_Drivers/timer.hpp

```diff
@@ -0,0 +1,157 @@
+#pragma once
+#include <cstdint>
+#include "L0_LowLevel/LPC40xx.h"
+#include "L0_LowLevel/system_controller.hpp"
+#include "L1_Drivers/pin.hpp"
+#include "utility/log.hpp"
+
+
+class TimerInterface
+{
+ public:
+  inline static LPC_TIM_TypeDef * tim_register[4];
+  inline static LPC_SC_TypeDef * sc_ptr = LPC_SC;
+  enum RegisterChoice : uint8_t
+  {
+    kRegMR0 = 0,
+    kRegMR1 = 1,
+    kRegMR2 = 2,
+    kRegMR3 = 3,
+    kRegCR0 = 4,
+    kRegCR1 = 5
+  };
+  enum TimerPort : uint8_t
+  {
+    kTimer0 = 0,
+    kTimer1 = 1,
+    kTimer2 = 2,
+    kTimer3 = 3
+  };
+  // Modes for each match register that when match register matches timer
+  // system can trigger an Interrupt, Stop Counter, or Restart and in any
+  // mixture of those commands.
+  enum TimerSetting : uint8_t
+  {
+    kInterrupt            = 0b001,
+    kRestart              = 0b010,
+    kStop                 = 0b100,
+    kInterruptRestart     = 0b011,
+    kInterruptStop        = 0b101,
+    kRestartStop          = 0b110,
+    kInterruptRestartStop = 0b111
+  };
+  virtual void Initialize(TimerPort timer)= 0;
+  virtual void SetTimer(uint32_t time, RegisterChoice reg,
+                        TimerSetting mode) = 0;
+  virtual uint32_t GetTimer() = 0;
+  virtual void ClearInterrupts() = 0;
+};
+
+class Timer : public TimerInterface, protected Lpc40xxSystemController
+{
+ public:
+  static constexpr uint8_t kTime0Mat0   = 0b011;
+  static constexpr uint8_t kClearMode   = 0b0111;
+  static constexpr uint8_t kPrescaleMicro   = 48;
```

Comment:
What if the user chooses to change the system frequency to something else? This makes the assumption that the cpu speed is 48 MHz which may not be the case. Use `GetPeripheralFrequency()` in place of where you use the PreScalar. Also allow the user to select a prescaler rather than defining it for them. I would like to be able to control the prescalar's value. Let the developer figure out how to make a 1ms timer rather than trying to figure that out here.